### PR TITLE
Add opt-in strict string format validation

### PR DESCRIPTION
**Pull request by zzstoatzz**
adds pydantic validation per https://github.com/MarshalX/atproto/issues/406#issuecomment-2488334102. see the [usage example](https://github.com/MarshalX/atproto/pull/451#discussion_r1855078671)

| format | example | pattern/constraints |
|--------|---------|---------|
| `did` | `did:plc:z72i7hdynmk6r22z27h6tvur` | `did:[a-z]+:[a-zA-Z0-9._%-]+` (max 2KB, no /?#[]@) |
| `handle` | `alice.bsky.social` | Domain name: 2+ segments, ASCII alphanumeric/hyphens, 1-63 chars per segment, max 253 total. Last segment no leading digit. |
| `at-uri` | `at://alice.bsky.social/app.bsky.feed.post/3jxtb5w2hkt2m` | `at://` + handle/DID + optional `/collection/rkey`. Max 8KB. No query/fragment in Lexicon. |
| `datetime` | `2024-11-24T06:02:00Z` | ISO 8601/RFC 3339 with required 'T', seconds, timezone (Z/±HH:MM). No -00:00. |
| `nsid` | `app.bsky.feed.post` | 3+ segments: reversed domain (lowercase alphanum+hyphen) + name (letters only). Max 317 chars. |
| `tid` | `3jxtb5w2hkt2m` | 13 chars of [2-7a-z]. First byte's high bit (0x40) must be 0. |
| `record-key` | `3jxtb5w2hkt2m` | 1-512 chars of [A-Za-z0-9._:~-]. "." and ".." forbidden. |
| `uri` | `https://example.com/path` | RFC-3986 URI with letter scheme + netloc/path/query/fragment. Max 8KB, no spaces. |
| `did:plc` | `did:plc:z72i7hdynmk6r22z27h6tvur` | Method identifier must be 24 chars of base32 ([a-z2-7]). Cannot include underscore. |

does not address open questions regarding the [language](https://github.com/MarshalX/atproto/blob/main/packages/atproto_client/models/languages.py) types or `re2`. happy to chat on those if they'd be blocking here

may close https://github.com/MarshalX/atproto/issues/406

### notes

I ran this to generate models
```
uvx poetry run atproto gen models
```

and this to update docs
```
uvx poetry run make -s -C docs gen
```

and this to run the new tests
```
uvx poetry run pytest tests/test_atproto_client/models/tests/test_string_formats.py
```

if desirable, I can follow up this PR to add contribution instructions to the current README section (for `uv` at least, and perhaps a TODO for proper `poetry` usage, since I don't really use that) - let me know!

---

in/valid examples for tests are sourced from https://github.com/bluesky-social/atproto/tree/main/interop-test-files/syntax

### CODE REVIEW

File: tests/test_atproto_client/models/tests/test_string_formats.py
Comment (zzstoatzz) [inline tests/test_atproto_client/models/tests/test_string_formats.py:1]: let me know:

- if you know a good place to go grab some representative data for these

- if I should add tests elsewhere covering generated interfaces

File: examples/advanced_usage/validate_string_formats.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+assert validated_good_handle == some_good_handle  # noqa: S101
+
+print(f'{validated_good_handle=}\n\n')
+
+try:
+    print('Trying to validate a bad handle with strict validation...')
+    HandleTypeAdapter.validate_python(some_bad_handle, context=strict_validation_context)
+except ValidationError as e:
+    print(e)
```
Comment (zzstoatzz) [inline examples/advanced_usage/validate_string_formats.py:30]: just left this breadcrumb here in case its helpful, happy to remove it!

File: packages/atproto_client/models/string_formats.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+def only_validate_if_strict(validate_fn: core_schema.WithInfoValidatorFunction) -> Callable:
+    """Skip validation if not opting into strict validation."""
+
+    def wrapper(v: str, info: ValidationInfo) -> str:
+        if not (info and isinstance(info.context, Mapping) and info.context.get(_OPT_IN_KEY, False)):
+            return v
+        return validate_fn(v, info)
+
+    return wrapper
```
Comment (zzstoatzz) [inline packages/atproto_client/models/string_formats.py:41]: here's the critical "pydantic usage" making use of [validation context](https://docs.pydantic.dev/latest/concepts/validators/#validation-context)

File: pyproject.toml [hunk 3]
```diff
@@ -143,17 +162,25 @@ ignore = [
...
+    "D104",
+    "D100",
+    "D107",
+    "D103",
+    "D415", # missing docstring
+    "D101",
+    "D102", # missing docstring in public class and method
 ]
-"tests/*.py" = ["S101", "D"]
+"tests/**.py" = ["S101", "D", "ANN201"]
```
Comment (zzstoatzz) [inline pyproject.toml:183]: applies ruff ignores over all python files in tests, adding ignores of "no return annotation" - happy to remove in favor of explicit `-> None`

see [ANN201](https://docs.astral.sh/ruff/rules/missing-return-type-undocumented-public-function/)

File: packages/atproto_codegen/models/generator.py [hunk 4]
```diff
@@ -161,6 +162,25 @@ def _get_str_enum_typehint(nsid: NSID, field_type_def: models.LexString, *, opti
...
+            'handle': 'string_formats.Handle',
+            'nsid': 'string_formats.Nsid',
+            'tid': 'string_formats.Tid',
+            'record-key': 'string_formats.RecordKey',
+            'uri': 'string_formats.Uri',
+            'language': 'string_formats.Language',
+        }
+        str_typehint = format_map.get(format_type, 'str')
+
+    # Handle existing enum/known_values logic
```
Comment (zzstoatzz) [inline packages/atproto_codegen/models/generator.py:183]: im just pattern matching here, let me know if I have the right idea

File: packages/atproto_codegen/models/generator.py [hunk 5]
```diff
@@ -139,7 +140,7 @@ def _get_optional_typehint(type_hint: str, *, optional: bool) -> str:
 
 
 def _get_str_enum_typehint(nsid: NSID, field_type_def: models.LexString, *, optional: bool) -> str:
-    values = field_type_def.known_values or field_type_def.enum
+    values = field_type_def.known_values or field_type_def.enum or []
```
[the remark this answers is not in the document]
Comment (zzstoatzz) [reply packages/atproto_codegen/models/generator.py:143]: i can revert it buts its because both variables in this expression are optional, which appears to come right from `LexString` field definitions
```python
field_type_def.known_values or field_type_def.enum
```

File: packages/atproto_codegen/models/generator.py [hunk 6]
```diff
@@ -63,13 +63,14 @@ def save_code_part(nsid: NSID, code: str) -> None:
 
 
 def _get_model_imports() -> str:
-    # we are using ruff with F401 autofix to delete unused imports
```
[the remark this answers is not in the document]
Comment (zzstoatzz) [reply packages/atproto_codegen/models/generator.py:66]: oops! will add this back

File: pyproject.toml [hunk 7]
```diff
@@ -167,3 +194,6 @@ ignore = [
 docstring-quotes = "double"
 multiline-quotes = "double"
 inline-quotes = "single"
+
+[tool.pytest.ini_options]
+asyncio_default_fixture_loop_scope = "session"
```
[the remark this answers is not in the document]
Comment (zzstoatzz) [reply pyproject.toml]: its just because pytest complains if you dont, I reverted this for now in [6ad722a](https://github.com/MarshalX/atproto/pull/451/commits/6ad722a5964483d3cc00d601c59db7126990b658)

**Comment by zzstoatzz** [comment]
thanks for the review @MarshalX! comments should be addressed in [6ad722a](https://github.com/MarshalX/atproto/pull/451/commits/6ad722a5964483d3cc00d601c59db7126990b658)

File: tests/test_atproto_client/models/tests/test_string_formats.py
Comment (MarshalX) [reply tests/test_atproto_client/models/tests/test_string_formats.py:1]: @zzstoatzz hi pls use this files for tests https://github.com/bluesky-social/atproto/tree/main/interop-test-files/syntax. Just download it from the official repo and put to this one. pls do not rename of modify this files. because we want to keep it with easy sync with the official repo without additional effort

File: packages/atproto_client/models/string_formats.py (hunk 2 above)
Comment (MarshalX) [reply packages/atproto_client/models/string_formats.py:41]: could we try to invert this if statement? to avoid using of "not"

smth like:

```suggestion
def only_validate_if_strict(validate_fn: core_schema.WithInfoValidatorFunction) -> Callable:
    """Skip validation if not opting into strict validation."""

    def wrapper(v: str, info: ValidationInfo) -> str:
        if info and isinstance(info.context, Mapping) and info.context.get(_OPT_IN_KEY, False):
            return validate_fn(v, info)
        return v

    return wrapper
```

Comment (zzstoatzz) [reply packages/atproto_client/models/string_formats.py]: aha, good call. updated

File: tests/test_atproto_client/models/tests/test_string_formats.py
Comment (zzstoatzz) [reply tests/test_atproto_client/models/tests/test_string_formats.py:1]: sounds good! i copied the directory wholesale from the repo you linked into `tests/test_atproto_client/interop-test-files/*` and then updated the tests to read from that

File: tests/test_atproto_client/models/tests/test_string_formats.py [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+        'record_key': valid_record_keys[0],
+        'uri': 'https://example.com',  # No interop test file for URI
+        'language': 'en-US',  # No interop test file for language
+    }
+
+
+@pytest.fixture
+def invalid_data() -> dict:
+    return {
+        'handle': get_test_cases('handle_syntax_invalid.txt')[0],
```
Comment (MarshalX) [inline tests/test_atproto_client/models/tests/test_string_formats.py:87]: could we parametrize it as well? to use all the test cases from .txt files instead of the first one

File: tests/test_atproto_client/models/tests/test_string_formats.py
Comment (MarshalX) [reply tests/test_atproto_client/models/tests/test_string_formats.py:1]: great! i left one comment. so basically it is about use all the cases from these files instead of the only one per each

File: tests/test_atproto_client/models/tests/test_string_formats.py [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+    assert instance.did == valid_data['did']
+
+    # Test invalid handle fails
+    try:
+        get_or_create(
+            {'handle': invalid_data['handle'], 'did': valid_data['did']}, TestModel, strict_string_format=True
+        )
+        pytest.fail('Handle validation should have failed')
+    except ModelError as e:
+        assert 'must be a domain name' in str(e)
```
Comment (MarshalX) [inline tests/test_atproto_client/models/tests/test_string_formats.py:181]: pytest has syntax sugar to not deal with exceptions in try-except blocks. check this one:

https://github.com/MarshalX/atproto/blob/8aafc9a4019b0a6cd68da73a27e5f4355225d102/tests/test_atproto_client/models/tests/test_changed_lexicon_compatability.py#L71-L80

it will automatically handle if exception is not occurred and will properly fail the test

File: packages/atproto_client/models/string_formats.py [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+    r'[a-zA-Z]'  # Last segment must start with letter
+    r'[a-zA-Z0-9-]*'  # Middle chars
+    r'[a-zA-Z]'  # Must end with letter
+    r'$'  # End of string
+)
+LANG_RE = re.compile(r'^(i|[a-z]{2,3})(-[A-Za-z0-9-]+)?$')
+RKEY_RE = re.compile(r'^[A-Za-z0-9._:~-]{1,512}$')
+TID_RE = re.compile(rf'^[2-7a-z]{{{TID_LENGTH}}}$')
+CID_RE = re.compile(r'^[A-Za-z0-9+]{8,}$')
+AT_URI_RE = re.compile(
```
Comment (MarshalX) [inline packages/atproto_client/models/string_formats.py:43]: sorry for the long response! i am busy at work :(

as I can see this regex is different to official one: 

https://github.com/bluesky-social/atproto/blob/53fcc2fbcbda9b325eeffb5705126831baf2a427/packages/syntax/src/aturi.ts#L5)

We use official in our AtUri class:

https://github.com/MarshalX/atproto/blob/e011bb4c18e85967a90a0271ea384972d1ae8292/packages/atproto_core/uri/uri.py#L9

Probably, your regex come from re-impl mentioned in the issues. But it would be awesome to align with official one! And as I can see you can get everything from the `syntax` package! https://github.com/bluesky-social/atproto/tree/53fcc2fbcbda9b325eeffb5705126831baf2a427/packages/syntax/src. With it all the tests should pass and no need for excluding test cases

Comment (zzstoatzz) [reply packages/atproto_client/models/string_formats.py:43]: > sorry for the long response! i am busy at work :(

no worries! so am I 🙂 

ack! I missed that - thank you!

I will update the regexs to follow those as soon as I can 👍

**Comment by zzstoatzz** [comment]
hi @MarshalX - ok all of the cases are passing with no skipping!

I found that the spec and test cases in the interop valid/invalid `.txt` examples were not entirely covered by the types in `atproto_core`, so for now I have tended towards duplicating regex in `string_formats.py` and documenting with comments why I included the logic I did.

For example, looking at the AT-URI validation, I noticed what appear to be few differences between the regex I added and `atproto_core`'s `AtUri` that seem important for passing the interop tests:

1. The spec requires "at://" prefix (see tests like "a://did:plc:asdf123" and "at:/did:plc:asdf123" being invalid)
2. The path must be a valid NSID format (invalid like "at://did:plc:asdf123/12345" and "at://did:plc:asdf123/short/stuff" show this)
3. The handle authority section has some idiosyncrasy - tests fail cases like "at://name", "at://name.0", "at://bsky" that a general URI parser might accept

let me know if you think i'm missing something here or if you think I should move some of these requirements into the types in `atproto_core` instead

also, just in case it comes up: i had a tricky problem with loading the test cases where I was calling `.strip()` on each line of the `.txt` files, but one of the datetime test cases was checking for whitespace 🙃

File: packages/atproto_client/models/utils.py [hunk 11]
```diff
@@ -92,13 +97,18 @@ def _get_or_create(
...
+            strict=strict,
+            strict_string_format=strict_string_format,
+        )
 
     if isinstance(model_data, model):
         return model_data
 
     try:
-        return model(**model_data)
+        return model.model_validate(model_data, context={'strict_string_format': strict_string_format})
```
Comment (MarshalX) [inline packages/atproto_client/models/utils.py:111]: We could not replace constructor to the `model_validate` here without additional checks. For example, it break firehose client. Because `get_or_create` could take to `model` arg not only pydantic`s models, but also python`s data classes! So, to be able to pass `context` we should add `isinstance` with check that this is pydantic model

File: tests/test_atproto_client/models/tests/test_string_formats.py [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+
+
+@pytest.fixture
+def valid_tids() -> List[str]:
+    return get_test_cases('tid_syntax_valid.txt')
+
+
+@pytest.fixture
+def valid_record_keys() -> List[str]:
+    return get_test_cases('recordkey_syntax_valid.txt')
```
Comment (MarshalX) [inline tests/test_atproto_client/models/tests/test_string_formats.py:95]: should we delete these fixtures? as I can see we are using it only once in the `valid_data` fixture. and for `invalid_data` we have different approach. we call directly `get_test_cases` without fixture per each. so i want to pick 1 way to do the same (i prefer not to create a lot of fixtures per each file)

File: tests/test_atproto_client/models/tests/test_string_formats.py [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+        get_or_create({'handle': invalid_data['handle'], 'did': valid_data['did']}, FooModel, strict_string_format=True)
+        pytest.fail('Handle validation should have failed')
+    except ModelError as e:
+        assert 'must be a domain name' in str(e)
+
+    # Test invalid did fails
+    try:
+        get_or_create({'handle': valid_data['handle'], 'did': invalid_data['did']}, FooModel, strict_string_format=True)
+        pytest.fail('Did validation should have failed')
+    except ModelError as e:
```
Comment (MarshalX) [inline tests/test_atproto_client/models/tests/test_string_formats.py:212]: could we do the same here as in `test_string_format_validation` test?

like:
```python
with pytest.raises(ModelError) as exc_info:
    get_or_create(blabla)

assert 'must be in blabla' in str(exc_info.value)
```

File: packages/atproto_client/models/string_formats.py [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+    def wrapper(v: str, info: ValidationInfo) -> str:
+        if info and isinstance(info.context, Mapping) and info.context.get(_OPT_IN_KEY, False):
+            return validate_fn(v, info)
+        return v
+
+    return wrapper
+
+
+@only_validate_if_strict
+def validate_handle(v: str, info: ValidationInfo) -> str:
```
Comment (MarshalX) [inline packages/atproto_client/models/string_formats.py:66]: as i can see the `info` arg is not used? could be names like `_` instead

File: packages/atproto_client/models/string_formats.py [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+Language = Annotated[str, BeforeValidator(validate_language)]
+RecordKey = Annotated[str, BeforeValidator(validate_record_key)]
+Cid = Annotated[str, BeforeValidator(validate_cid)]
+AtUri = Annotated[str, BeforeValidator(validate_at_uri)]
+DateTime = Annotated[str, BeforeValidator(validate_datetime)]  # see pydantic-extra-types #239
+Tid = Annotated[str, BeforeValidator(validate_tid)]
+Uri = Annotated[str, BeforeValidator(validate_uri)]
+
+# Any valid ATProto string format
+ATProtoString = Annotated[
```
Comment (MarshalX) [inline packages/atproto_client/models/string_formats.py:215]: `AtProtoString` instead? (lower `t`)

**Review by MarshalX — COMMENTED** [review]
such an amazing hard work you did! let's move that discussion about different of the `AtUri` to different issue; for now we are okay to have 2 validation logic.

 i have nothing to add except comments that i left and question about documentation. since we support `.format` of the `LexString` do we want to show it in the documentation? Right now fields are not show their format at the atproto.blue. We could create separate issue for it to finally merge your great PR

The link to built docs from this PR is https://atproto--451.org.readthedocs.build/en/451/

File: tests/test_atproto_client/models/tests/test_string_formats.py (hunk 12 above)
Comment (zzstoatzz) [reply tests/test_atproto_client/models/tests/test_string_formats.py]: ahh my fault for being sloppy there - updated to isolate all the io and removed all unused fixtures

File: packages/atproto_client/models/string_formats.py (hunk 14 above)
Comment (zzstoatzz) [reply packages/atproto_client/models/string_formats.py]: i updated `only_validate_if_strict` to selectively pass `info` - let me know what you think!

File: packages/atproto_client/models/utils.py (hunk 11 above)
Comment (zzstoatzz) [reply packages/atproto_client/models/utils.py]: i see - I updated the checks like [this](https://github.com/zzstoatzz/atproto/blob/strict-string-format/packages/atproto_client/models/utils.py#L110-L117) - what do you think?

**Comment by zzstoatzz** [comment]
> let's move that discussion about different of the AtUri to different issue

> We could create separate issue for it to finally merge your great PR

will do! thanks for all the help thus far 🙂

File: packages/atproto_client/models/string_formats.py (hunk 14 above)
Comment (MarshalX) [reply packages/atproto_client/models/string_formats.py]: to be honest i am not sure about it. since there is no use case for use of 'info' the logic to support it is complicated and uses inspect module :( could we just drop this arg? without checking needs or not

Comment (zzstoatzz) [reply packages/atproto_client/models/string_formats.py]: makes sense! i was thinking its likely one would want access to `info` later, but no need to get ahead of ourselves!

I've simplified it in [bad5512](https://github.com/MarshalX/atproto/pull/451/commits/bad5512e3bc8b8c7996d81ecdc56f7876aab5ea4)

**Comment by MarshalX** [comment]
i did the last check before merging and found something around TypedDict. could you pls check? probably we should not use new strict string formats for TypedDict?

image

**Comment by zzstoatzz** [comment]
hi @MarshalX - hrmm can you share more on how you ended up with that? I may be missing something silly but can't directly reproduce that (with an example like this)

<details>

<summary>example</summary>

```python
from typing import TypedDict

from atproto_client.models import base, string_formats
from pydantic import TypeAdapter, ValidationError


class Params(base.ParamsModelBase):
    actor: string_formats.Handle


class ParamsDict(TypedDict):
    actor: string_formats.Handle


# Test data
some_good_handle = 'test.bsky.social'
some_bad_handle = 'invalid@ @handle'

# Test both with and without strict validation
strict_context = {'strict_string_format': True}
non_strict_context = {'strict_string_format': False}

# Initialize adapters with explicit types
HandleTypeAdapter: TypeAdapter[str] = TypeAdapter(string_formats.Handle)
ParamsTypeAdapter: TypeAdapter[ParamsDict] = TypeAdapter(ParamsDict)
ParamsModelAdapter: TypeAdapter[Params] = TypeAdapter(Params)


def run_validation_test(context: dict[str, bool]) -> None:
    mode = 'strict' if context['strict_string_format'] else 'non-strict'
    print(f'\n=== Testing {mode} mode ===')

    print('\n1. Direct Handle validation:')
    try:
        handle_result: str = HandleTypeAdapter.validate_python(some_bad_handle, context=context)
        print(f'Validation passed: {handle_result=}')
    except ValidationError as e:
        print(f'Validation failed: {e!s}')

    print('\n2. TypedDict validation:')
    try:
        params_result: ParamsDict = ParamsTypeAdapter.validate_python({'actor': some_bad_handle}, context=context)
        print(f'Validation passed: {params_result=}')
    except ValidationError as e:
        print(f'Validation failed: {e!s}')

    print('\n3. Pydantic Model validation:')
    try:
        model_result: Params = ParamsModelAdapter.validate_python({'actor': some_bad_handle}, context=context)
        print(f'Validation passed: {model_result=}')
    except ValidationError as e:
        print(f'Validation failed: {e!s}')


run_validation_test(non_strict_context)
run_validation_test(strict_context)
```

</details>


<details>

<summary>stdout</summary>

```python
» python examples/advanced_usage/validate_string_formats.py

=== Testing non-strict mode ===

1. Direct Handle validation:
Validation passed: handle_result='invalid@ @handle'

2. TypedDict validation:
Validation passed: params_result={'actor': 'invalid@ @handle'}

3. Pydantic Model validation:
Validation passed: model_result=Params(actor='invalid@ @handle')

=== Testing strict mode ===

1. Direct Handle validation:
Validation failed: 1 validation error for function-before[wrapper(), str]
  Value error, Invalid handle: must be a domain name (e.g. user.bsky.social) with max length 253 [type=value_error, input_value='invalid@ @handle', input_type=str]
    For further information visit https://errors.pydantic.dev/2.10/v/value_error

2. TypedDict validation:
Validation failed: 1 validation error for typed-dict
actor
  Value error, Invalid handle: must be a domain name (e.g. user.bsky.social) with max length 253 [type=value_error, input_value='invalid@ @handle', input_type=str]
    For further information visit https://errors.pydantic.dev/2.10/v/value_error

3. Pydantic Model validation:
Validation failed: 1 validation error for Params
actor
  Value error, Invalid handle: must be a domain name (e.g. user.bsky.social) with max length 253 [type=value_error, input_value='invalid@ @handle', input_type=str]
    For further information visit https://errors.pydantic.dev/2.10/v/value_error
```

</details>

**Comment by MarshalX** [comment]
@zzstoatzz i mean does it rly need to be complex type with pydantic validation for TypedDict? Speaking of the highlighted problem on the screenshot, probably PyCharm (Intellij + Python plugin in my case) is just dumb xD

I do not say that it won't work. It will. But IDE`s analyzers struggle with it. 

We use TypedDict only to help the users with field suggestions in IDE. And for static type analyzers (for someone who cares). When the user uses dict-based params instead of proper model. But under the hood, we will use proper model. So this is just a syntax sugar for the users who want to write it fast in unsafe dict-driven Python manure

tl;dr no problems with it; we just do not use TypedDict directly (only for type annotating); my IDE could not handle it

**Review by MarshalX — APPROVED** [review]

**Comment by zzstoatzz** [comment]
oh ok! I was using pylance strict but I'm happy to take a stab at any further type-smoothing that would be helpful!

thanks again for all your help 🙂
